pkg/kubelet/winstats/perfcounter_nodestats.go

```diff
@@ -0,0 +1,151 @@
+// +build windows
+
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package winstats
+
+import (
+	"errors"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+	"github.com/lxn/win"
+)
+
+// PerfCounterNodeStatsClient is a client that provides Windows Stats via PerfCounters
+type PerfCounterNodeStatsClient struct {
+	nodeStats
+	mu sync.RWMutex
+}
+
+// NewPerfCounterNodeStatsClient creates a new client
+func NewPerfCounterNodeStatsClient() *PerfCounterNodeStatsClient {
+	return &PerfCounterNodeStatsClient{}
+}
+
+func (p *PerfCounterNodeStatsClient) startMonitoring() error {
+	memory, err := getPhysicallyInstalledSystemMemoryBytes()
+	if err != nil {
+		return err
+	}
+	p.nodeStats.memoryPhysicalCapacityBytes = memory
+
+	version, err := exec.Command("cmd", "/C", "ver").Output()
+	if err != nil {
+		return err
+	}
+	p.kernelVersion = strings.TrimSpace(string(version))
+
+	cpuChan, err := readPerformanceCounter(cpuQuery)
+	if err != nil {
+		return err
+	}
+
+	memWorkingSetChan, err := readPerformanceCounter(memoryPrivWorkingSetQuery)
+	if err != nil {
+		return err
+	}
+
+	memCommittedBytesChan, err := readPerformanceCounter(memoryCommittedBytesQuery)
+	if err != nil {
+		return err
+	}
+
+	go p.startNodeMonitoring(cpuChan, memWorkingSetChan, memCommittedBytesChan)
+	return nil
+}
+
+func (p *PerfCounterNodeStatsClient) getMachineInfo() (*cadvisorapi.MachineInfo, error) {
```

Comment:
All your `get*()` functions essentially access the `nodeStat`. You need to acquire the lock.
Alternatively, you can always call `getNodeStats()` internally if you need to access any field in the `nodeStats`

Reply:
I believe the only get function that needs locking is `getNodeStats()` because the node stats (cpu, memory, etc) are updated from a different goroutine. However, all the get* functions (`getMachineInfo()`, `getVersionInfo()`) don't access any state from another goroutine, so I don't think they need a lock...